fix(api,app): Block protocol start if the Flex Stacker is not homed or shuttle is missing + home stacker from Desktop app.

vegano1 opened the pull request:
# Overview

The Flex stacker Platform might not always be at the homed (gripper) position, which could cause a problem when loading labware on a stacker occupied column 4, since the labware would not be physically where we expected. So let's ensure we can't start the protocol if the relevant stackers are not homed or the shuttle is missing. The Platform missing status coming from the backend was also wrong, let's also fix that.

## Stacker needs home
<img width="929" alt="Screenshot 2025-06-24 at 11 18 35 AM" src="https://github.com/user-attachments/assets/5d276fc9-b364-4d99-b138-0cd36afd1387" />

## Stacker Missing shuttle

<img width="915" alt="Screenshot 2025-06-24 at 12 09 31 PM" src="https://github.com/user-attachments/assets/2fb9629e-ba16-469a-9170-c622c9b7ebdd" />

## Module needs setup

<img width="925" alt="Screenshot 2025-06-24 at 10 13 14 AM" src="https://github.com/user-attachments/assets/084c83f5-34ab-4fa2-af60-da2772eb2b34" />

## Modules and fixtures ready

<img width="920" alt="Screenshot 2025-06-24 at 12 09 57 PM" src="https://github.com/user-attachments/assets/0fa68c6b-ba7c-455e-90b8-156cae8ac986" />


Closes: [EXEC-1563](https://opentrons.atlassian.net/browse/EXEC-1563) [EXEC-1572](https://opentrons.atlassian.net/browse/EXEC-1572)

## Test Plan and Hands on Testing

- [x] Make sure the "Play" or "Start Protocol" button is disabled on the ODD and Desktop app if a stacker needs homing or the shuttle is missing.
- [x] Make sure that the "Home Stacker" button is shown in protocol setup for the Desktop App and ODD if the stacker needs to be homed.
- [x] Make sure the "Action needed" button is shown in protocol setup on the desktop app if the shuttle is missing
- [x] Make sure the "missing-shuttle" status is shown in protocol setup on the ODD if the shuttle is missing

## Changelog

- Move `PlatformState.MISSING` logic to the `get_platform_sensor_state` method of the FlexStacker Reader, since we need up-to-date limit switch status
- Added 'set_refresh_state' to the FlexStacker module in the API to tell the reader when to refetch states not polled, like limit switch status, TOF sensors, etc.
- Change "Calibration needed" copy in deck configuration to "Action needed" per design
- Added "Home Stacker" button to the protocol setup of the desktop app if the flex stacker needs to home.
- Block protocol start if the flex stacker is not homed or the platform is not extended

## Review requests

- Makes sense?

## Risk assessment

Low, unreleased.

[EXEC-1563]: https://opentrons.atlassian.net/browse/EXEC-1563?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[EXEC-1572]: https://opentrons.atlassian.net/browse/EXEC-1572?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

vegano1 committed:
fix(api,app): Block protocol start if the Flex Stacker is not homed or shuttle is missing.

sfoster1 commented on app/src/organisms/Desktop/Devices/ProtocolRun/SetupModuleAndDeck/SetupModulesList.tsx:
```diff
@@ -255,12 +284,31 @@ export function ModulesListItem({
     attachedModuleMatch.moduleOffset?.last_modified == null
 
-  if (needsCalibration || stackerShuttleMissing) {
+  if (stackerNeedsHome) {
```
can we put this section below the `needsCalibration || stackerShuttleMissing` one and then lose the logic checks in the `disabled` state in line 294 and the disabled tooltip in ll298-303?

sfoster1 commented on app/src/resources/runs/useModuleCalibrationStatus.ts:
```diff
@@ -39,9 +38,16 @@ export function useModuleCalibrationStatus(
-      m => m.attachedModuleMatch?.moduleOffset?.last_modified == null
-    )
+    moduleData.some(m => {
+      if (m.attachedModuleMatch?.moduleType === FLEX_STACKER_MODULE_TYPE) {
```
this is probably not the right place fo rthis, since it will disable the run button with "calibrate attached module(s) to continue" - wants its own reason, right?

vegano1 replied:
@smb2268 and I spoke about it,

I think it should be, I think the confusion is that calibration is now a part of the greater "Module Setup" flow. However, some areas reference "calibration required," which should say "Module Setup Required" or something similar.

vegano1 replied:
Spoke to design, will change the copy to "Action needed".

sfoster1 reviewed (approved):
Approving to unblock since this is  time sensitive, but there's some small things I think we can improve

vegano1 committed:
change "Calibration needed" copy in deck configuration to "Action needed" per design
move PlatformState.MISSING logic into `get_platform_sensor_state` method of the FlexStacker Reader since we need up-to-date limit switch status

vegano1 committed:
show missing shuttle status

ahiuchingau reviewed (approved):
Tested on robot and worked as expected

vegano1 committed:
lint
